File: backend/app/api/schemas/matching.py
"""Request/response schemas for match links (WP-6).

Three shapes, and the split is about who is reading.

`MatchSummary` is what a **session** or a **planned session** carries about its
own link: enough for a badge and a link to the other side, and nothing that
costs a second query. It hangs off both read shapes, so neither resource is
silent about a state the athlete can see on the calendar.

`MatchRead` is the link **as its own resource** — the score, its whole
breakdown, and enough of both sides to render a proposal inbox row without
fetching either. A proposal is answered from a list, and a list that made the
client fetch two resources per row to say "your 2 h ride on Tuesday, against
the 90 min endurance session planned for Monday" is a list nobody can answer
from.

`SessionMatchState` is what the operations that *remove* a link answer with:
the session's status and whatever link (if any) now stands. Rejecting a
proposal is not a session edit, and answering with the whole session — its
recordings, its metric artefact — would be a payload the client did not ask
for on every button press.
"""

import datetime as dt
import uuid

from pydantic import BaseModel, ConfigDict

from app.api.pagination import Page
from app.domain.activity import SessionDiscipline, SessionMatchStatus
from app.domain.athlete import Discipline
from app.domain.matching import MatchComponent, MatchLinkStatus
from app.domain.purpose import Purpose
from app.domain.sessions import SessionStatus


class MatchComponentRead(BaseModel):
    """One component of a similarity score, and what it compared.

    The two raw numbers travel with the ratio because the ratio alone is not
    explicable: 0.62 on duration means nothing until it reads "90 min
    prescribed against 56 min ridden".
    """

    component: MatchComponent
    #: Agreement, in ``[0, 1]``.
    score: float
    #: The weight actually applied — the nominal weight renormalised over the
    #: components that could be assessed, so the applied weights sum to 1.
    weight: float
    #: The weight the build plan gives this component, before renormalisation.
    nominal_weight: float
    #: The prescribed quantity: seconds, watts or bpm, or a count of units.
    planned: float
    #: Its recorded counterpart, in the same unit.
    actual: float
    #: Which channel or unit the pair is in (``power``, ``hr``, ``intervals``,
    #: ``sets``); null for duration, which has only seconds.
    basis: str | None


class MatchUnassessedRead(BaseModel):
    """One component that had nothing to compare, and why.

    Not an error and not a zero: the weights renormalise over what remains, so
    this is the record of *what the score was made of*. A similarity of 0.9
    over two components is a different claim from the same number over three.
    """

    component: MatchComponent
    nominal_weight: float
    reason: str


class MatchBreakdownRead(BaseModel):
    """How a similarity score was arrived at."""

    #: The score itself. **Null is not zero**: it means no component could be
    #: assessed at all, which is why the link is a proposal rather than a
    #: refusal.
    score: float | None
    #: The nominal weights, by component — the constants the build plan fixes.
    weights: dict[MatchComponent, float]
    components: list[MatchComponentRead]
    not_assessed: list[MatchUnassessedRead]


class MatchSummary(BaseModel):
    """One link, as the two resources it joins carry it."""

    id: uuid.UUID
    session_id: uuid.UUID
    planned_session_id: uuid.UUID
    status: MatchLinkStatus
    #: Null when nothing could be compared; see `MatchBreakdownRead.score`.
    similarity: float | None
    #: When the athlete made this link their own. Null while it is still a
    #: machine verdict the athlete has not ruled on.
    confirmed_at: dt.datetime | None


class MatchSessionContext(BaseModel):
    """Enough of the completed session to render a proposal row."""

    id: uuid.UUID
    local_date: dt.date
    discipline: SessionDiscipline
    status: SessionMatchStatus
    #: Recording time for a device session, wall clock for a typed-in one —
    #: the same number the session list shows.
    duration_s: float


class MatchPlannedContext(BaseModel):
    """Enough of the planned session to render a proposal row."""

    id: uuid.UUID
    date: dt.date
    discipline: Discipline
    purpose: Purpose
    status: SessionStatus
    #: The athlete's own one-line intent, when they wrote one.
    intent_text: str | None


class MatchRead(MatchSummary):
    """One link as its own resource, with both sides and the whole breakdown."""

    breakdown: MatchBreakdownRead
    #: `app.domain.actor.Actor` in string form — ``system`` for a link matching
    #: made on ingest, ``athlete`` for one made by hand, ``agent:<label>`` for
    #: one an agent proposed.
    created_by: str
    #: What this link displaced, and therefore what unlinking restores.
    previous_session_status: SessionMatchStatus
    previous_planned_status: SessionStatus
    session: MatchSessionContext
    planned_session: MatchPlannedContext
    created_at: dt.datetime
    updated_at: dt.datetime


MatchesPage = Page[MatchRead]


class SessionMatchState(BaseModel):
    """Where a completed session stands relative to the plan, after an edit."""

    session_id: uuid.UUID
    status: SessionMatchStatus
    #: The link that now stands, or null when none does.
    match: MatchSummary | None


class MatchOutcomeRead(SessionMatchState):
    """What one run of matching decided (`POST /sessions/{id}/rematch`)."""

    #: How many planned sessions were in the window and still unlinked.
    candidates: int
    #: True when an existing confirmed or displaced link was left alone. The
    #: run decided nothing, deliberately: those are the athlete's own words
    #: and no re-run overwrites them.
    sticky: bool


class MatchCreate(BaseModel):
    """Payload for linking a session to a planned session by hand."""

    model_config = ConfigDict(extra="forbid")

    session_id: uuid.UUID
    planned_session_id: uuid.UUID
    #: Set when the athlete trained and it was **not** this: the planned
    #: session becomes ``displaced`` rather than completed, and the activity is
    #: scored standalone with no adherence axes (build plan WP-6.4).
    displaced: bool = False


class MatchRetarget(BaseModel):
    """Payload for pointing an existing link at a different planned session."""

    model_config = ConfigDict(extra="forbid")

    planned_session_id: uuid.UUID


class SessionMerge(BaseModel):
    """Payload for folding a second recording of one ride into this session."""

    model_config = ConfigDict(extra="forbid")

    #: The session whose recordings move onto this one. That session row is
    #: removed; its recordings and their stream files are kept.
    absorbed_session_id: uuid.UUID

File: frontend/lib/matching.ts
import type { components } from "@/generated/api/schema";
import { formatDurationClock } from "@/lib/format";

type Schemas = components["schemas"];

export type MatchLinkStatus = Schemas["MatchLinkStatus"];
export type MatchSummary = Schemas["MatchSummary"];
export type Match = Schemas["MatchRead"];
export type MatchBreakdown = Schemas["MatchBreakdownRead"];
export type MatchComponent = Schemas["MatchComponent"];
export type MatchComponentScore = Schemas["MatchComponentRead"];
export type MatchUnassessed = Schemas["MatchUnassessedRead"];
export type PlannedSessionListItem = Schemas["PlannedSessionListItem"];

/**
 * Every cached match list, however it was filtered — and, separately, every
 * cached single match.
 *
 * Two constants rather than one, because a query key is matched by **prefix**
 * over its elements and `"/api/v1/matches/{match_id}"` is a different second
 * element from `"/api/v1/matches"`, not a longer key beginning with it.
 * Invalidating the list alone leaves the panel reading a link whose status the
 * click just changed — which looks exactly like a mutation that silently did
 * nothing.
 */
export const MATCHES_QUERY_PREFIX = ["get", "/api/v1/matches"] as const;
export const MATCH_QUERY_PREFIX = [
  "get",
  "/api/v1/matches/{match_id}",
] as const;
/** Every cached page of the plan, which a link changes the status of. */
export const PLANNED_SESSIONS_QUERY_PREFIX = [
  "get",
  "/api/v1/planned-sessions",
] as const;
/**
 * One planned session by id — a different cache from the list above, for the
 * reason spelled out for matches: the second element is a different string,
 * not a longer one. Today's panel and the calendar's session sheet both read
 * it, so anything that rewrites a prescription has to name it explicitly.
 */
export const PLANNED_SESSION_QUERY_PREFIX = [
  "get",
  "/api/v1/planned-sessions/{planned_session_id}",
] as const;
/** The calendar week, whose cards carry the link state (WP-6, WP-7). */
export const PLAN_WEEK_QUERY_PREFIX = ["get", "/api/v1/plan/week"] as const;

/**
 * What one link claims, said in English.
 *
 * Keyed by the generated enum so a member added to `MatchLinkStatus` is a
 * compile error here rather than a badge rendering `auto_high` at an athlete.
 *
 * The pair worth reading carefully is `auto_high` against `confirmed`: both
 * mean the session and the planned session are joined, and they differ in
 * *whose* claim it is. A machine link may be revised by the next re-run of
 * matching; the athlete's never is (`STICKY_STATUSES`).
 */
export const MATCH_LINK_LABELS: Readonly<Record<MatchLinkStatus, string>> = {
  auto_high: "Auto-linked",
  pending: "Proposed",
  confirmed: "Confirmed",
  displaced: "Instead of",
};

/** Why a link says what it says, on hover and to a screen reader. */
export const MATCH_LINK_REASONS: Readonly<Record<MatchLinkStatus, string>> = {
  auto_high:
    "Linked without asking, because the recording and the prescription agree closely. Still yours to undo.",
  pending:
    "A proposal waiting on you: arc thinks this session answered that planned session, and has changed nothing until you say so",
  confirmed:
    "You linked these yourself, and no re-run of matching will revise it",
  displaced:
    "You trained, and deliberately not this: the planned session counts as displaced rather than missed or completed, and this session is scored on its own",
};

/**
 * The thresholds `app.domain.matching.classify` applies, mirrored.
 *
 * Mirrored rather than fetched because they are constants of the build plan
 * (WP-6.3) and not settings — the API states no endpoint for them, and a
 * number the UI can only get by reading the backend source is exactly the
 * number to write down beside the sentence it justifies. Nothing here decides
 * a link: they are used only to *explain* a score the API already classified,
 * which is why a drift between them and the domain misphrases a sentence
 * rather than mislabelling a link.
 */
export const AUTO_LINK_SIMILARITY = 0.75;
export const PROPOSAL_SIMILARITY = 0.4;

/** The three things a similarity is made of, named. */
export const COMPONENT_LABELS: Readonly<Record<MatchComponent, string>> = {
  duration: "Duration",
  intensity: "Intensity",
  structure: "Structure",
};

/** What each component compares, for the athlete who has not read WP-6. */
export const COMPONENT_DESCRIPTIONS: Readonly<Record<MatchComponent, string>> =
  {
    duration: "Prescribed seconds against recorded seconds",
    intensity:
      "Prescribed normalized power (or heart rate) against what was recorded",
    structure: "Prescribed efforts or sets against those detected or logged",
  };

/**
 * One side of one component, in the unit that component is measured in.
 *
 * The raw numbers travel with every ratio because a ratio alone is not
 * explicable (`MatchComponentRead`): 0.38 on duration means nothing until it
 * reads "57:00 prescribed against 2:29:00 ridden".
 */
export function formatComponentValue(
  component: MatchComponent,
  basis: string | null,
  value: number,
): string {
  if (component === "duration") {
    return formatDurationClock(value);
  }
  if (component === "intensity") {
    return basis === "hr"
      ? `${Math.round(value)} bpm`
      : `${Math.round(value)} W`;
  }
  const unit = basis === "sets" ? "sets" : "efforts";
  return `${Math.round(value)} ${unit}`;
}

/** A similarity as a whole percentage: `0.6872` → `69%`. */
export function formatSimilarity(score: number): string {
  return `${Math.round(score * 100)}%`;
}

/** A weight as a whole percentage, for the column that says how much it counted. */
export function formatWeight(weight: number): string {
  return `${Math.round(weight * 100)}%`;
}

/**
 * Why a score is null, said as the API means it.
 *
 * Null is not zero (`MatchBreakdownRead.score`): it means no component could
 * be assessed at all, which is why the link is a question rather than a
 * refusal.
 */
export const NO_SCORE_REASON =
  "Nothing could be compared: the date and the discipline agree, and none of the three components had two sides to put against each other";

/**
 * How a score compares to the thresholds, in a sentence.
 *
 * Said rather than shown as a gauge because the number is only meaningful
 * against the two cut-offs it was classified by, and a bar with no marks on it
 * would invite reading 0.69 as "nearly right" when what it means is "inside
 * the band where arc asks instead of deciding".
 */
export function describeScore(score: number | null): string {
  if (score === null) {
    return NO_SCORE_REASON;
  }
  if (score >= AUTO_LINK_SIMILARITY) {
    return `At or above ${formatSimilarity(AUTO_LINK_SIMILARITY)}, arc links a session without asking.`;
  }
  if (score >= PROPOSAL_SIMILARITY) {
    return `Between ${formatSimilarity(PROPOSAL_SIMILARITY)} and ${formatSimilarity(AUTO_LINK_SIMILARITY)}, arc proposes and leaves the decision to you.`;
  }
  return `Below ${formatSimilarity(PROPOSAL_SIMILARITY)}, arc proposes nothing on its own — this link is yours.`;
}

/** Whether a score is low enough that "done instead of" is the likelier claim. */
export function isDisplacementScore(score: number | null): boolean {
  return score !== null && score < PROPOSAL_SIMILARITY;
}

/**
 * What linking as *displaced* actually does, in plain language.
 *
 * WP-6.4 in one sentence, and it is not a detail: the difference between the
 * two link kinds is what the week says about the planned session afterwards
 * and whether the ride is scored against a prescription it never followed.
 */
export const DISPLACED_EXPLANATION =
  "Use this when you trained, but not the thing on the calendar. The planned session is marked displaced — neither missed nor completed — and this session is scored on its own terms, with no adherence to a prescription it never followed.";

/** The same, for the ordinary link. */
export const CONFIRMED_EXPLANATION =
  "Use this when this session is the planned one, however far the numbers drifted. The planned session is marked completed and this session is scored against its prescription.";
